ci/build_common.sh

```diff
@@ -14,6 +14,9 @@ CUDA_ARCHS= # Empty, use presets by default.
 GLOBAL_CMAKE_OPTIONS=()
 DISABLE_CUB_BENCHMARKS= # Enable to force-disable building CUB benchmarks.
 
+GLOBAL_CMAKE_OPTIONS+=("-DCMAKE_CUDA_FLAGS=-DCCCL_SUPPRESS_ICC_DEPRECATION_WARNING")
```

Comment:
@alliepiper is this the right place to add these? Or would it be better in the presets?

Reply:
I wondered about a few places now. @miscco suggested the top-level `CMakeLists.txt` yesterday. I also considered any of the cmake configs for CUB, Thrust and libcu++. Or the presets.

However, I want everybody who is using ICC to get this warning. So if an ICC user uses presets, they should get the warning. Same when including CCCL as sub_directory and turning on tests (so not suppression from CMake in a test build), etc. I only want to suppress the warning in the CI, which is why I chose the CI scripts. Some users may still use those to test locally, so an even better place would be inside the GitHub action, but that is scary territory for me.

I also wonder what @alliepiper has to say about this!

Reply:
For example here: https://github.com/NVIDIA/cccl/blob/main/cub/cub/cmake/cub-config.cmake#L92-L109. Which can be set by a cmake option (default `OFF`) here: https://github.com/NVIDIA/cccl/blob/main/cub/CMakeLists.txt#L54-L57.

I could extend those places (and find the right place for libcu++ as well), but it actually feels quite invasive for us wanting users to move away from the compiler anyway.

Reply:
> We have some flags for disabling other warnings in our builds already (like C++11 deprecation warnings). Where are those?

The C++11 suppressions are enabled in the presets, where they're only defined for the relevant builds. For example, https://github.com/NVIDIA/cccl/blob/main/CMakePresets.json#L209

> For example here: https://github.com/NVIDIA/cccl/blob/main/cub/cub/cmake/cub-config.cmake#L92-L109. Which can be set by a cmake option (default OFF) here: https://github.com/NVIDIA/cccl/blob/main/cub/CMakeLists.txt#L54-L57.
>
> I could extend those places (and find the right place for libcu++ as well), but it actually feels quite invasive for us wanting users to move away from the compiler anyway.

Defining the cmake option (defaulting to OFF) will have no impact on users and will make it easier to disable the warnings when we need to, since we can just set a CMake option directly instead of injecting through `CMAKE_CUDA_FLAGS`. I'd recommend following the existing patterns for the other deprecation suppressions: adding a CMake option that adds the compiler definition, and then enabling that option on the relevant builds.

The other deprecation suppression flags follow the convention `[PROJ]_IGNORE_DEPRECATED_[THING]`, so I'd suggest renaming the compiler def and CMake option to `CCCL_IGNORE_DEPRECATED_ICC`.

> is this the right place to add these? Or would it be better in the presets?

I think the base preset would be fine, or in the build scripts when we detect ICC. The current implementation is problematic because it's going through `-DCMAKE_CUDA_FLAGS` instead of setting a CMake option directly -- if we need to modify `CMAKE_CUDA_FLAGS` using the `ci/build_foo.sh -cmake-options ....` syntax, the separate definitions of `CMAKE_CUDA_FLAGS` will overwrite each other. This will also cause issues on the non-CUDA thrust configurations, since those will read `CMAKE_CXX_FLAGS` instead of `CMAKE_CUDA_FLAGS`.

We can also set the CMake option in the matrix.yaml file, e.g.

```
    - {jobs: ['build'], std: 'all', cxx: ['intel'], cmake_options: '-DCCCL_IGNORE_DEPRECATED_ICC=ON'}
```

This has the advantage of only disabling the warning for the specific cases where we want it off.

Personally, I think adding it to the `base` preset would be the easiest path forward.

Reply:
It might be time to start consolidating some of the redundant compiler flag logic into a CCCL compiler interface, similar to how we do for Thrust/CUB ([example](https://github.com/NVIDIA/cccl/blob/main/cub/cmake/CubBuildCompilerTargets.cmake)).

Reply:
> It might be time to start consolidating some of the redundant compiler flag logic into a CCCL compiler interface, similar to how we do for Thrust/CUB ([example](https://github.com/NVIDIA/cccl/blob/main/cub/cmake/CubBuildCompilerTargets.cmake)).

I totally agree! I could follow your suggestions above for Thrust and CUB, but the cmake scripts for libcu++ are different. I think having a shared CCCL target that Thrust, CUB and libcu++ link against would greatly simplify my effort. Any chance you could give that a shot? :) Thx!

Reply:
Given that we have consolidated our cmake infrastructure I would rather not have a cmake option but do the right thing in the cmake infrastructure

There is no case where we do not want this in CI and also there is no case where we want a user to toggle this